### PR TITLE
Fixed mypy errors

### LINKED ISSUE #909: [BUG] MyPy test failures on main
Looks like mypy is failing on the current main. I am not sure what changed / how this got into main unless it was just caused by a update to mypy itself.

Example: this PR which changed nothing about the code https://github.com/microsoft/graspologic/pull/908
Test output (unit-and-doc-test): 
```{pytb}
Run mypy ./graspologic
graspologic\match\qap.py:420: error: Argument "key" to "sorted" has incompatible type "Callable[[Union[Sequence[Sequence[Sequence[Sequence[Sequence[Any]]]]], _SupportsArray[dtype[Any]], Sequence[_SupportsArray[dtype[Any]]], Sequence[Sequence[_SupportsArray[dtype[Any]]]], Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]], Sequence[Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]]], complex, Sequence[Union[complex, str, bytes]], Sequence[Sequence[Union[complex, str, bytes]]], Sequence[Sequence[Sequence[Union[complex, str, bytes]]]], Sequence[Sequence[Sequence[Sequence[Union[complex, str, bytes]]]]]]], Union[Sequence[Sequence[Sequence[Sequence[Any]]]], Any, _SupportsArray[dtype[Any]], Sequence[_SupportsArray[dtype[Any]]], Sequence[Sequence[_SupportsArray[dtype[Any]]]], Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]], complex, Sequence[Union[complex, str, bytes]], Sequence[Sequence[Union[complex, str, bytes]]], Sequence[Sequence[Sequence[Union[complex, str, bytes]]]]]]"; expected "Callable[[Union[Sequence[Sequence[Sequence[Sequence[Sequence[Any]]]]], _SupportsArray[dtype[Any]], Sequence[_SupportsArray[dtype[Any]]], Sequence[Sequence[_SupportsArray[dtype[Any]]]], Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]], Sequence[Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]]], complex, Sequence[Union[complex, str, bytes]], Sequence[Sequence[Union[complex, str, bytes]]], Sequence[Sequence[Sequence[Union[complex, str, bytes]]]], Sequence[Sequence[Sequence[Sequence[Union[complex, str, bytes]]]]]]], Union[SupportsDunderLT, SupportsDunderGT]]"
graspologic\match\qap.py:420: error: Value of type "Union[Sequence[Sequence[Sequence[Sequence[Sequence[Any]]]]], _SupportsArray[dtype[Any]], Sequence[_SupportsArray[dtype[Any]]], Sequence[Sequence[_SupportsArray[dtype[Any]]]], Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]], Sequence[Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]]], complex, Sequence[Union[complex, str, bytes]], Sequence[Sequence[Union[complex, str, bytes]]], Sequence[Sequence[Sequence[Union[complex, str, bytes]]]], Sequence[Sequence[Sequence[Sequence[Union[complex, str, bytes]]]]]]" is not indexable
graspologic\match\qap.py:420: error: Incompatible return value type (got "Union[Sequence[Sequence[Sequence[Sequence[Any]]]], Any, _SupportsArray[dtype[Any]], Sequence[_SupportsArray[dtype[Any]]], Sequence[Sequence[_SupportsArray[dtype[Any]]]], Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]], complex, Sequence[Union[complex, str, bytes]], Sequence[Sequence[Union[complex, str, bytes]]], Sequence[Sequence[Sequence[Union[complex, str, bytes]]]]]", expected "Union[SupportsDunderLT, SupportsDunderGT]")
graspologic\match\qap.py:433: error: Item "int" of "Union[int, RandomState, Generator, None]" has no attribute "permutation"
graspologic\match\qap.py:433: error: Item "None" of "Union[int, RandomState, Generator, None]" has no attribute "permutation"
graspologic\match\qap.py:457: error: Item "int" of "Union[int, RandomState, Generator, None]" has no attribute "uniform"
graspologic\match\qap.py:457: error: Item "None" of "Union[int, RandomState, Generator, None]" has no attribute "uniform"
graspologic\cluster\autogmm.py:559: error: Incompatible types in assignment (expression has type "List[Optional[RandomState]]", variable has type "ndarray[Any, dtype[signedinteger[Any]]]")
graspologic\embed\base.py:439: error: <nothing> has no attribute "ndim"
graspologic\embed\base.py:441: error: <nothing> has no attribute "ndim"
graspologic\embed\base.py:444: error: <nothing> has no attribute "shape"
graspologic\embed\base.py:446: error: <nothing> has no attribute "shape"
graspologic\embed\base.py:449: error: Incompatible types in assignment (expression has type "Union[ndarray[Any, Any], Any]", variable has type "List[Union[ndarray[Any, Any], Any]]")
graspologic\embed\case.py:205: error: Incompatible types in assignment (expression has type "floating[Any]", variable has type "float")
Found 14 errors in 4 files (checked 79 source files)
Error: Process completed with exit code 1.
```

Side note: when I get the latest version of mypy and run it locally on `main`, I actually get 
```
Found 393 errors in 55 files (checked 78 source files)
``` 

I wont post the whole traceback but these errors involve basically every module in the library. Many are private functions.

**A few larger points for discussion**: 
- Would anyone be opposed to taking mypy out of unit-and-doc-test and putting it as it's own check, much like black has its own code-format-check? It seems unnecessary to check on every os/version like we do for actual tests.
- Can someone explain to me what mypy gives us that beartype doesn't? I know it is static and beartype is dynamic, right? but in practice what is this doing for us? 
   - If nothing, can we remove? 
- What is our policy on mypy compatibility for legacy code that is private? 
- any tips on getting mypy to tell me what the function is that is actually causing the error? --verbose did not do it for me.

### PR DESCRIPTION
**Pull request by bdpedigo**
- [x] Does this PR have a descriptive title that could go in our release notes?
- [ ] Does this PR add any new dependencies?
- [ ] Does this PR modify any existing APIs?
   - [ ] Is the change to the API backwards compatible?
- [ ] Have you built the documentation (reference and/or tutorial) and verified the generated documentation is appropriate?

#### Reference Issues/PRs
Fixes #909

#### Other changes
- `black` just updated to 22.1.0 which is now stable. There was a change to whitespace around power operators (see  changelog https://black.readthedocs.io/en/stable/change_log.html#id1)

#### Notes
- This does NOT address #910 which is also blocking anything going into `dev`

### CODE REVIEW

File: graspologic/match/qap.py [hunk 1]
```diff
@@ -416,9 +416,8 @@ def _quadratic_assignment_faq(
     # check outlier cases
     if n == 0 or partial_match_value.shape[0] == n:
         # Cannot assume partial_match is sorted.
-        partial_match_value = np.row_stack(
-            sorted(partial_match_value, key=lambda x: x[0])
-        )
+        sort_inds = np.argsort(partial_match_value[:, 0])
```
Comment (bdpedigo) [inline graspologic/match/qap.py:419]: @asaadeldin11 if you have the time (if not no worries): 1. does this fix seem equivalent to you and 2. was there a test for this edge case anywhere? I remember this coming up but dont remember if it had a test associated

File: graspologic/embed/case.py [hunk 2]
```diff
@@ -202,7 +202,7 @@ def _get_tuning_parameter(
 
         # just use the ratio of the leading eigenvalues for the
         # tuning parameter, or the closest value in its possible range.
-        self.alpha_ = np.float_(L_top / YYt_top)
```
Comment (bdpedigo) [inline graspologic/embed/case.py:205]: @loftusa use of `np.float_` was causing issues with mypy - if there is some reason you think this call is necessary, can you propose an alternative fix ASAP?

Comment (loftusa) [reply graspologic/embed/case.py:205]: I think if I called it, there may have been a reason, but I'm not sure what it was at this point. I'd be down to just change it to a normal `float` call and see if anything happens

**Comment by bdpedigo** [comment]
hi @pbourke, @daxpryce mentioned you might be the person to talk to about mypy issues. I was able to resolve most of the errors that popped up, but these last two are stumping me:

1)
This one comes from [this line](https://github.com/microsoft/graspologic/blob/2b13f047d1d312cce0a9efc82f1298e180c21507/graspologic/match/qap.py#L579) where I had to pull in a private function from scipy and add some typing
```pytb
graspologic/match/qap.py:594: error: Argument 1 to "RandomState" has incompatible type "Union[<subclass of 
"RandomState" and "Integral">, <subclass of "RandomState" and "integer">, <subclass of "Generator" and "Integral">, 
<subclass of "Generator" and "integer">]"; expected "Union[None, Union[Union[_SupportsArray[dtype[Union[bool_, 
integer[Any]]]], Sequence[_SupportsArray[dtype[Union[bool_, integer[Any]]]]], 
Sequence[Sequence[_SupportsArray[dtype[Union[bool_, integer[Any]]]]]], 
Sequence[Sequence[Sequence[_SupportsArray[dtype[Union[bool_, integer[Any]]]]]]], 
Sequence[Sequence[Sequence[Sequence[_SupportsArray[dtype[Union[bool_, integer[Any]]]]]]]]], Union[bool, int, 
Sequence[Union[bool, int]], Sequence[Sequence[Union[bool, int]]], Sequence[Sequence[Sequence[Union[bool, int]]]], 
Sequence[Sequence[Sequence[Sequence[Union[bool, int]]]]]]], BitGenerator]"
```

2) This one comes from [this line](https://github.com/microsoft/graspologic/blob/a8048f1631f2f0676555cc3f04796bf6053a6496/graspologic/embed/base.py#L449) in some type checking for a base class
```pytb
graspologic/embed/base.py:449: error: Incompatible types in assignment (expression has type "Union[ndarray[Any, Any], 
Any]", variable has type "List[Union[ndarray[Any, Any], Any]]")
```

Any tips would be appreciated as these errors are blocking anything from being merged into dev at the moment

File: graspologic/embed/case.py (hunk 2 above)
Comment (bdpedigo) [reply graspologic/embed/case.py:205]: and I am guessing both of those start as float anyway? regardless feel free to propose a fix and test that it doesnt mess up mypy if you think the change should be something else

Comment (loftusa) [reply graspologic/embed/case.py:205]: Woops, didn't see this was a PR.
Yeah, this change should be fine

**Comment by daxpryce** [comment]
> This one comes from [this line](https://github.com/microsoft/graspologic/blob/2b13f047d1d312cce0a9efc82f1298e180c21507/graspologic/match/qap.py#L579) where I had to pull in a private function from scipy and add some typing
> 
> ```
> graspologic/match/qap.py:594: error: Argument 1 to "RandomState" has incompatible type "Union[<subclass of 
> "RandomState" and "Integral">, <subclass of "RandomState" and "integer">, <subclass of "Generator" and "Integral">, 
> <subclass of "Generator" and "integer">]"; expected "Union[None, Union[Union[_SupportsArray[dtype[Union[bool_, 
> integer[Any]]]], Sequence[_SupportsArray[dtype[Union[bool_, integer[Any]]]]], 
> Sequence[Sequence[_SupportsArray[dtype[Union[bool_, integer[Any]]]]]], 
> Sequence[Sequence[Sequence[_SupportsArray[dtype[Union[bool_, integer[Any]]]]]]], 
> Sequence[Sequence[Sequence[Sequence[_SupportsArray[dtype[Union[bool_, integer[Any]]]]]]]]], Union[bool, int, 
> Sequence[Union[bool, int]], Sequence[Sequence[Union[bool, int]]], Sequence[Sequence[Sequence[Union[bool, int]]]], 
> Sequence[Sequence[Sequence[Sequence[Union[bool, int]]]]]]], BitGenerator]"

So, to start with, don't make a copy of the function - there are these things called type stubs (in a file extension of .pyi) where you can describe the type hint on the function you're calling from that library.  That's actually how typeshed works in general!

https://github.com/python/typeshed/blob/master/stubs/dateparser/dateparser/date_parser.pyi as a very short example.

Notice that the folder structure matches up, but the only thing in this file is the type hint.  

> 2. This one comes from [this line](https://github.com/microsoft/graspologic/blob/a8048f1631f2f0676555cc3f04796bf6053a6496/graspologic/embed/base.py#L449) in some type checking for a base class
> 
> ```
> graspologic/embed/base.py:449: error: Incompatible types in assignment (expression has type "Union[ndarray[Any, Any], 
> Any]", variable has type "List[Union[ndarray[Any, Any], Any]]")
> ```
> 
> Any tips would be appreciated as these errors are blocking anything from being merged into dev at the moment

ndarrays aren't generecized - it's just List[Union[ndarray, Any]], which itself turns into `List[Any]`.  Are you sure you want to say "give me a list of ndarrays and/or anything you want"?

Also, there's a difference between `Union[List[ndarray, List[List[Union[float, int]]]]` and `List[Union[ndarray, List[List[Union[float, int]]]]`.  One requires all of the items be ndarrays xor a list of lists, the other requires a list of either.

**Comment by bdpedigo** [comment]
> So, to start with, don't make a copy of the function

I thought in general you were averse to relying on private functions from other libraries? That was my main reason for pulling it in actually. If you disagree I can go the type stub route

**Comment by bdpedigo** [comment]
so since everything but 3.9 is passing, I assert that I've fixed #909 (with help from @daxpryce) but not #910.

**Comment by bdpedigo** [comment]
Beat my head against this for a while to get this to work on my machine to no avail (I basically tried a bunch of different versions of numpy)

A few related threads:
https://stackoverflow.com/questions/66060487/valueerror-numpy-ndarray-size-changed-may-indicate-binary-incompatibility-exp

https://stackoverflow.com/questions/40845304/runtimewarning-numpy-dtype-size-changed-may-indicate-binary-incompatibility

https://stackoverflow.com/questions/68478456/solved-error-numpy-ndarray-size-changed-may-indicate-binary-incompatibility

several people mention something about installing with `--no-binary` when using pip but that didn't seem to work for me either, also not sure if that could even be a solution for all users. Lots of these threads also mention something about `pycocotools` but I have not seen that library in any of my installs.

**Comment by bdpedigo** [comment]
closing in favor of #913
